dbtc/client/cloud/base.py

```diff
@@ -968,46 +975,172 @@ def test_connection(self, account_id: int, payload: Dict) -> Dict:
+
         """
-        return self._simple_request(
+
+        def run_status_formatted(run: Dict, time: float) -> str:
+            """Format a string indicating status of job.
+            Args:
+                run (dict): Dictionary representation of a Run
+                time (float): Elapsed time since job triggered
+            """
+            status = JobRunStatus(run['data']['status']).name
+            url = run['data']['href']
+            try:
+                step = run['data']['run_steps'][-1]
+                current_step = f'{step["index"]}. {step["name"]}'
+            except IndexError:
+                current_step = 'N/A'
+            return (
+                f'Status: "{status.capitalize()}", Elapsed time: {round(time, 0)}s'
+                f', Current step: "{current_step}", View here: {url}'
+            )
+
+        def step_meets_condition(
+            step: Dict,
+            command_type: str,
+            statuses: List[str],
+        ) -> bool:
+            return (
+                any(command in step['name'] for command in COMMAND_TYPES[command_type])
+                and step['status_humanized'].lower() in statuses
+            )
+
+        if restart_from_failure:
+            self.console.log(f'Restarting job {job_id} from last failed state.')
+            last_run_data = self.list_runs(
+                account_id=account_id,
+                include_related=['run_steps'],
+                job_definition_id=job_id,
+                order_by='-id',
+                limit=1,
+            )['data'][0]
+
+            last_run_status = last_run_data['status_humanized'].lower()
+            last_run_id = last_run_data['id']
+
+            if last_run_status == 'error':
+                rerun_steps = []
+
+                for run_step in last_run_data['run_steps']:
+
+                    # get the dbt command used within this step
+                    command = run_step['name'].partition('`')[2].partition('`')[0]
+                    if not any(cmd for cmd in ALL_COMMANDS if cmd in command):
+                        self.console.log(
+                            f'Skipping rerun for command "{run_step["name"]}" '
+                            'as it does not need to be repeated.'
+                        )
+                    else:
+
+                        # skipped and other commands should be rerun entirely
+                        if step_meets_condition(
+                            run_step, 'run', ['skipped']
+                        ) or step_meets_condition(
+                            run_step, 'other', ['error', 'skipped', 'failed']
+                        ):
+                            rerun_steps.append(command)
+
+                        # errors and failures are when we need to inspect to figure
+                        # out the point of failure
+                        elif step_meets_condition(run_step, 'run', ['error', 'failed']):
```

Comment:
@dpguthrie this is looking like a solid step forward! One other thing occurred to me: `--vars`. For non-run (e.g. anything other than `dbt build` and `dbt run`) and skipped steps, it looks like this is copying the `--vars` statements correctly. But on failed steps like `dbt test` or the point of failure step, those extra arguments are being dropped.

Noting also that I don't think we need to worry about replicating any `--selectors`, `--exclude` etc. because those statements will scope the initial run, and we're just working within that scope for the restart from failure.